home/tests/unit/dailywalk/test_get.py

```diff
@@ -0,0 +1,187 @@
+        # Create a user
+        response = self.client.post(
+            path="/api/appuser/create",
+            data={
+                "name": "Abhay Kashyap",
+                "email": "abhay@blah.com",
+                "zip": "72185",
+                "age": 99,
+                "account_id": "12345",
+            },
+            content_type="application/json",
+        )
+
+        # Check for a successful response by the server
+        self.assertEqual(response.status_code, 200)
+        # Parse the response
+        response_data = response.json()
+        fail_message = f"Server response - {response_data}"
+        self.assertEqual(response_data["status"], "success", msg=fail_message)
+        self.assertEqual(response_data["message"], "App User registered successfully", msg=fail_message)
+
+        # Create three daily walks
+        self.daily_walks = [
+            {"account_id": "12345", "event_id": "777", "date": "2020-02-21", "steps": 1500},
+            {"account_id": "12345", "event_id": "888", "date": "2020-02-22", "steps": 500},
+            {"account_id": "12345", "event_id": "999", "date": "2020-02-23", "steps": 1000},
+        ]
+        # Create daily walks
+        for daily_walk in self.daily_walks:
+            response = self.client.post(path="/api/dailywalk/create", data=daily_walk, content_type="application/json",)
+            # Check for a successful response by the server
+            self.assertEqual(response.status_code, 200)
+            # Parse the response
+            response_data = response.json()
+            fail_message = f"Server response - {response_data}"
+            self.assertEqual(response_data["status"], "success", msg=fail_message)
+            self.assertEqual(response_data["message"], "Dailywalk recorded successfully", msg=fail_message)
+
+        # Details for Daily walk list view
+        self.url = "/api/dailywalk/get"
+        # Request parameters
+        self.request_params = {"account_id": "12345"}
+        # Content type
+        self.content_type = "application/json"
+
+    def test_dailwalk_get_failure(self):
+        self.request_params["account_id"] = "0000000"
+
+        # Send the request
+        response = self.client.post(path=self.url, data=self.request_params, content_type=self.content_type)
+        # Check for a successful response by the server
+        self.assertEqual(response.status_code, 200)
+        # Parse the response
+        response_data = response.json()
+        fail_message = f"Server response - {response_data}"
+        self.assertEqual(response_data["status"], "error", msg=fail_message)
+        self.assertEqual(
+            response_data["message"],
+            f'User does not exist for account - {self.request_params["account_id"]}. Please register first!',
+            msg=fail_message,
+        )
+
+    def test_dailwalk_get(self):
+        response = self.client.post(path=self.url, data=self.request_params, content_type=self.content_type)
```

Comment:
Should we test the case in which a user has two AppUser records with the same email...?

Reply:
Good call.. I haven't actually built the email aggregation and right now uses the id, but will update it and add the test

Reply:
I'm thinking that aggregation might be a little different between DailyWalk and IntentionalWalk records...

For IntentionalWalk, it can simply be the aggregation of records between AppUser accounts with the same email, as each record is a distinct explicit period recorded by the user.

However, for Daily Walk, there might be some overlap of days. I expect that, in order not to have gaps in the daily step data, on the client I'll probably always be querying some period back in time to ensure contiguous data, and so that the prior day's count is complete. That means that after signing into a new phone, the app may save some prior history back to the server that might overlap with the data from the previous device account. In this case, we can probably always just take the more recent record for a given day as its more likely to be correct.

Reply:
Not sure if I fully understand. But i can see daily walk being an issue across devices for a given day. More so if its not a switch to a new device but rather continuous usage across two devices. Off the top of my head, one way I can think of maintaining data consistency is that the app could query the server and look at the timestamp on the update (just added) and if its more recent than the last time the app data was refreshed, it can be changed accordingly and pushed back to the server. Basically, before everytime the daily walk call is made, you can make a get request to get fresh data and push it back along with the changes.. does this sound reasonable?

Reply:
Well, here's a hypothetical example. I wake up in the morning, take a walk, check the app, it saves the current day's steps so far back to the server.

In the afternoon, I pick up a new phone, download the app and sign in to it creating a new AppUser account, and at the end of the day it saves the day's total steps again back to the server.

So, there's the old AppUser with the partial day, and the new AppUser with the complete record. I don't think the app should have to query and check for existing records, the fact that there are duplicate accounts/records is a server-side implementation detail. The app and Apple Health/Google Fit, remain the source of truth, so whatever the latest data the app sends should be what the server maintains. If you want, you could delete the duplicate record from the old AppUser account at save time so that the aggregate query later is easier.

I'm not sure that there even needs to be a param to distinguish create/update of the daily walk data. The app will opportunistically just always send the latest daily walk data (going back a few days at least to prevent gaps) back to the server whenever the user opens the app (and maybe also on some sort of regular background wake-up schedule). And the server should just always save it with the most up-to-date data.

Although I'll be creating a single unique uuid key for the daily walk records for the local datastore which needs a single value primary key, the server could also just choose to ignore it and use the [AppUser, date] unique index like a compound primary key...

Reply:
Hmmm still not very sure how the data is recorded on the app and when its transmitted. 

> So, there's the old AppUser with the partial day, and the new AppUser with the complete record.

How would the new AppUser have the complete record? wouldn't it have the record for the remaining part of the day? 

> The app and Apple Health/Google Fit, remain the source of truth, so whatever the latest data the app sends should be what the server maintains.

So I guess I'm mainly unclear about this part.. Is the app accessing steps from these other apps like Google/Apple fit? and is it safe to assume that the data there will be synced across devices and hence, irrespective of the devices, the app will always have the total steps a person takes?

> I'm not sure that there even needs to be a param to distinguish create/update of the daily walk data.

Agreed. I could simply overwrite the record for a particular day if the app always has access to the total steps across device that is being fed in from a sync source of truth like google/apple fit

> The app will opportunistically just always send the latest daily walk data (going back a few days at least to prevent gaps) back to the server whenever the user opens the app (and maybe also on some sort of regular background wake-up schedule). And the server should just always save it with the most up-to-date data.

Unsure about when the updates are made. Would it be every time the app is opened? or would it be an update at a fixed time everyday? This would be important since in the former case, the app would need to figure out how many days of data to send. If we can figure this part out, I guess we can make the server handle things accordingly. The more generic use case would be if someone installs the app on 2 devices and uses them interchangeably which basically encompasses the use case where someone makes a switch. The simplest way would be to make sure the app checks when it last uploaded things to the server and then send steps for everyday since then. This will ensure edge cases like lack of internet or a phone thats turned off during a scheduled back-up time. This would also mean each device can upload its own steps. If they are synced and has aggregated data across devices from google/apple fit, the server can handle the logic of figuring out which record to use. If they aren't synced across apps and are being uploaded independently, then the server can aggregate things accordingly..

Reply:
Yes, that's exactly right- Apple Health (tied to Apple ID) and Google Fit (tied to Google account) on the respective platforms are the ultimate source of truth that the app queries.

As for the frequency of updates, that hasn't been decided yet. Certainly, it would be at least whenever the user opens the app. However, it should also be possible for each app to schedule a regular background "wakeup" schedule in which it queries Apple Health/Google Fit and sends the latest data. I know that Apple has started to restrict some of this background behavior in more recent iOS releases, so I need to double-check what is possible right now.

Where you have the edge case where the data is not automatically synced is: the user switches from iOS to Android or vice versa, or creates a new Apple/Google account with their new device.